Modules/Quotas/test/Quotas.Application.Tests/Tests/Repositories/MessagesRepositoryTests.cs

```diff
@@ -0,0 +1,146 @@
+        SystemTime.Set(halfOfMonth);
+
+        var messages = new List<MessageEntity>() {
+            CreateMessage(DateTime.Now, _identityAddress1),
+            CreateMessage(YESTERDAY, _identityAddress1),
+            CreateMessage(TOMORROW, _identityAddress1),
+            CreateMessage(LAST_YEAR, _identityAddress1),
+            CreateMessage(NEXT_YEAR, _identityAddress1)
+        };
+        await _messagesArrangeContext.Messages.AddRangeAsync(messages);
+        await _messagesArrangeContext.SaveChangesAsync();
+
+        var repository = new MessagesRepository(_actContext);
+        const QuotaPeriod quotaPeriod = QuotaPeriod.Month;
+
+        // Act
+        var count = await repository.Count(_identityAddress1, quotaPeriod.CalculateBegin(), quotaPeriod.CalculateEnd(), CancellationToken.None);
+
+        // Assert
+        count.Should().Be(3);
+    }
+
+    [Fact]
+    public async void Counts_entities_total_quotaPeriod()
+    {
+        // Arrange
+        var messages = new List<MessageEntity>() {
+            CreateMessage(DateTime.Now, _identityAddress1),
+            CreateMessage(TOMORROW, _identityAddress1),
+            CreateMessage(NEXT_YEAR, _identityAddress1)
+        };
+        await _messagesArrangeContext.Messages.AddRangeAsync(messages);
+        await _messagesArrangeContext.SaveChangesAsync();
+
+        var repository = new MessagesRepository(_actContext);
+        const QuotaPeriod quotaPeriod = QuotaPeriod.Total;
+
+        // Act
+        var count = await repository.Count(_identityAddress1, quotaPeriod.CalculateBegin(), quotaPeriod.CalculateEnd(), CancellationToken.None);
+
+        // Assert
+        count.Should().Be(3);
+    }
+
+
+    [Fact]
+    public async void Counts_entities_only_for_requested_identityAddress()
+    {
+        // Arrange
+        var messages = new List<MessageEntity>() {
+            CreateMessage(DateTime.Now, _identityAddress1),
+            CreateMessage(TOMORROW, _identityAddress2),
+            CreateMessage(NEXT_YEAR, _identityAddress1)
+        };
+        await _messagesArrangeContext.Messages.AddRangeAsync(messages);
+        await _messagesArrangeContext.SaveChangesAsync();
+
+        var repository = new MessagesRepository(_actContext);
+        const QuotaPeriod quotaPeriod = QuotaPeriod.Total;
+
+        // Act
+        var count = await repository.Count(_identityAddress1, quotaPeriod.CalculateBegin(), quotaPeriod.CalculateEnd(), CancellationToken.None);
+
+        // Assert
+        count.Should().Be(2);
+    }
+
+    private static MessageEntity CreateMessage(DateTime createdAt, IdentityAddress identityAddress)
+    {
+        return new(
+            identityAddress,
+            DeviceId.New(),
+            null,
+            Array.Empty<byte>(),
+            new List<Attachment>(),
+            new List<RecipientInformation>()
+            )
+        {
+            CreatedAt = createdAt
```

Comment:
You can get rid of the `InternalsVisibleTo` attribute and the setter of the `CreatedAt` property by changing the implementation of the `CreateMessage` method to something like this:

```csharp
SystemTime.Set(createdAt);

var message = new(
            identityAddress,
            DeviceId.New(),
            null,
            Array.Empty<byte>(),
            new List<Attachment>(),
            new List<RecipientInformation>()
            );

SystemTime.Reset();

return message;
```

Cases like this are the reason I introduced `SystemTime`.